hubble: Support for debug capture events

gandro opened the pull request:
This PR brings initial support for debug capture events. Ref #12416 

Most data printed by `cilium monitor -vv -t capture` should be available in the protobuf. This includes:

- The debug capture subtype
- The interface index and (if available) name (this is also added to trace events)
- The proxy port
- All flow information of the captured flow, including all the annotations that Hubble provides (e.g. destination names if DNS visibility is enabled etc)

The Hubble CLI needs to be updated to print all available information, but because capture events, it will be printed in `hubble observe -t capture` even in existing CLIs.

To be able to observe these events, one needs to enable debug events as documented here: https://docs.cilium.io/en/v1.9/operations/troubleshooting/#enabling-datapath-debug-messages As a follow-up, we probably also want to add CI for this.

The full proto output can be tested via:
```
grpcurl -plaintext -d '{"number":30,"whitelist":{"event_type":{"type": 3}}}' -unix /var/run/cilium/hubble.sock observer.Observer.GetFlows
``` 

Open questions:
- Do we really want to add the `proxy_port` field just for debug capture events, or should we maybe have a special field for datapath specific debug information (which could also include the datapath hash etc)?

Best reviewed per commit.

gandro committed:
hubble/parser: Initial support for DebugCapture events

This commits adds very basic support for debug capture events. None of
its fields are parsed yet, besides the type, subtype and captured
network package.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
api/hubble: Add DebugCapturePoint to Flow

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
hubble/parser: Decode DebugCapture subtype

This translates the debug capture subtype into a protobuf enum, similar
to the existing TraceObservationPoint or DropReason enums.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
pkg/datapath/link: Add GetIfNameCached

This commit adds a helper to map the ifindex to the interface name. It
is copied over from the monitor package and exported as a public
function, such that in can be accessed outside of the monitor package.

This commit contains no functional change.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
api/hubble: Add network interface to Flow message

This adds a new field to a flow, indicating the network interface on
which the flow was observed. A network interface has a numeric non-zero
index and an optional name.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
hubble/parser: Populate network interface for trace and capture events

This commit populates the network interface on which a trace or debug
catpure event was observed. This information is provided by the datapath
for certain trace points.

The interface index is always non-zero, so we treat zero as absent
information in the parser.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
api/hubble: Add proxy port field to flow

This is intended for debug capture messages.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
hubble/parser: Decode proxy port for debug capture events

This parses the proxy port for debug capture events and stores it in the
decoded flow.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

glibsm commented on pkg/hubble/parser/threefour/parser.go:
```diff
@@ -620,3 +622,30 @@ func decodeDebugCapturePoint(dbg *monitor.DebugCapture) pb.DebugCapturePoint {
+		return nil
+	}
+
+	name, _ := link.GetIfNameCached(ifIndex)
```
err ok to ignore here?

gandro replied:
Yes. I added a comment to indicate that it is intentional.

ti-mo commented on api/v1/flow/flow.proto:
```diff
@@ -291,6 +300,20 @@ enum TrafficDirection {
+    DBG_CAPTURE_POINT_UNKNOWN = 0;
+    reserved 1 to 3;
+    DBG_CAPTURE_DELIVERY = 4;
+    DBG_CAPTURE_FROM_LB =5;
```
ultra nit: missing space in `= 5`

ti-mo commented on pkg/datapath/link/link.go:
```diff
@@ -16,12 +16,25 @@ package link
+func init() {
+	go func() {
+		for {
+			linkCache.syncCache()
```
Since calling this involves a netlink call behind the scenes, what would be the intended failure mode? This will fail silently and cause the link cache to always be empty. Some form of error handling is likely required here.

I just noticed this was copied from `ifindex.go`, where errors also used to be silently dropped. In most cases, this call will fail due to permissions, so IMO this can exit or log error/crit.

gandro replied:
I don't think a failure here is critical enough to warrant killing the Cilium agent, but a logging the error and exiting the go routine makes a ton of sense, will add it.

ti-mo commented on pkg/hubble/parser/threefour/parser.go:
```diff
@@ -608,3 +620,50 @@ func getTCPFlags(tcp layers.TCP) string {
+		return nil
+	}
+
+	// if the interface it not found, name will be an empty string
```
Small typo here.

kaworu commented on pkg/datapath/link/link.go:
```diff
@@ -61,3 +74,40 @@ func GetIfIndex(ifName string) (uint32, error) {
+	defer linkCache.mu.RUnlock()
+
+	name, ok := linkCache.indexToName[ifIndex]
+	return name, ok
```
This probably want to use `c` instead of `linkCache` correct?

kaworu commented on pkg/hubble/parser/threefour/parser.go:
```diff
@@ -608,3 +620,50 @@ func getTCPFlags(tcp layers.TCP) string {
+}
+
+func decodeNetworkInterface(tn *monitor.TraceNotify, dbg *monitor.DebugCapture) *pb.NetworkInterface {
+	ifIndex := 0
```
nit: `ifIndex` could be `uint32` here, reducing a bit the casting dance following.

gandro committed:
pkg/datapath/link: Log error and exit in link cache

Listing the network links will likely fail due to permission errors. If
we observe such an error, we should log it and stop the cache sync
instead of silently ignoring it.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

tklauser commented on pkg/datapath/link/link.go:
```diff
@@ -61,3 +80,40 @@ func GetIfIndex(ifName string) (uint32, error) {
+}
+
+func (c *ifNameCache) syncCache() error {
+	indexToName := map[int]string{}
```
Minor nit: this could be preallocated with `len(links)` before the `for _, link := range links` loop below:

```go
indexToName := make(map[int]string, len(links))
```
